server/NewBank.java

```diff
@@ -3,35 +3,50 @@
 import java.util.HashMap;
 
 public class NewBank {
-	
 	private static final NewBank bank = new NewBank();
-	private HashMap<String,Customer> customers;
-	
+	/**
+	 * Created nested HashMap to hold the Customer, username, and password
+	 * relationships:
+	 * HashMap<userName, HashMap <Customer, String>> which maps to
+	 *  <String < Customer, password>>
+	 **/
+	private HashMap< String, HashMap <Customer, String> > customers;
+
 	private NewBank() {
 		customers = new HashMap<>();
+
 		addTestData();
 	}
-	
+
 	private void addTestData() {
 		Customer bhagy = new Customer();
 		bhagy.addAccount(new Account("Main", 1000.0));
-		customers.put("Bhagy", bhagy);
-		
+		/** put username, Customer and password into the HashMap for each customer */
+		customers.put("Bhagy", new HashMap(){{put(bhagy, "12345");}});
```

Comment:
Small suggestion, may be worth parameterising your HashMap instantiation instead of using the raw type. 

So `new HashMap<<Customer, String>(){{put(bhagy, "12345");}});` for example

Reply:
Actually, I've thought about it a bit more. Your current structure is to have a single username and password in a HashMap and then having a HashMap of HashMaps which may be a bit overkill. It works fine but alternate options to consider is:

Store the password in the Customer class as it's customer information like this:

`package newbank.server;

import java.util.ArrayList;

public class Customer {
	
	private ArrayList<Account> accounts;
	private String password;
	
	public Customer() {
		accounts = new ArrayList<>();
	}
	
	public String accountsToString() {
		String s = "";
		for(Account a : accounts) {
			s += a.toString();
		}
		return s;
	}

	public Boolean checkPassword(String password){
		return this.password.equals(password);
	}

	public void addAccount(Account account) {
		accounts.add(account);		
	}
}
`

Or change the customers HashMap to be a map of customerIDs and passwords instead of customerIDs and HashMap. This option is slightly worse in my opinion as it's handling Customer data in the NewBank class when NewBank shouldn't care about how Customers work internally.

Reply:
@ibrahimdirar I thought about this as an approach but opted for tweaking the HashMap in first instance. I do agree with your point about where things more clearly sit, so happy to make that change if others agree.  Should we pick up at the end of the first Sprint?

Reply:
Yes, that's seems the best forum to discuss

Reply:
Thanks - by the way, thanks for the feedback. Being new to this, I really appreciate it.

Reply:
I would like to add my proposition for this task. 

Analysing it, I have understood, that this task is more difficult than it looked like during the sprint preparation. 

1. What if we need a class Database which will keep (for the beginning) all Customers. 
- then the Database will have a HashMap of Customers (a list of all registered customers)
-  the Database will have such methods like: getAll, getById, getByPropName
- such simple API will help us to communicate with a single source of truth (DB) which will provide us al the data needed

2. What if we need Authorization service. (class Authorization)
 - which will have method login(username, password)
 - login will ask DB for a list of Customers, then will find a customer with a given username or (which is better) with a given id.
 - then it will compare passwords (no need to write details)
 - then "login" method will provide true or false (in the future it can provide a particular Customer object)
 
 Pros: 
 - we will bring more structure to this project. Moreover such simulation of a DB is already mentioned in the project (when we have hashMap of customers) but now, we will be able to scale our project and add to DB any other "tables" for instance "micro-loans" etc. So in the future, when we will be implementing micro-loans or any other more complex functionality, we will just add to our DB new HashMap(list).
 
 - in the future we may will need to add functionality like "register" "logout". If so, we will just add new methods to our Authorization service.

Reply:
I would suggest to split the current task on three parts:
1. Analysing the authorization approaches. (which is almost done)
2. Implementing basic DB api with a list of Customers, preparing a script for adding to DB mock data
3. Implement Authorization api for login and connect all the components, so the application will be able to login or reject the user.

Reply:
@MrDmitriev, thank you for reviewing. 

From my perspective, your logic makes sense and I agree we won’t get there in this sprint.  

When I looked at the challenge at the beginning of the week, I could see that this could be quite a complex implement, so I’m pleased that my attempt has prompted this thinking. I played with a number of implementation pathways, hitting dead ends as complexity grew, before settling on the nested HashMap for simplicity.   

All that said, I am now also wondering whether we should have invested some time in agreeing broad structure design principles up front, where we potentially would have arrived at this position before Sprint 1?  Not sure what the right balance is for agile development?  Planning up front or learn through doing??  

Given the requirements of the brief I agree that ultimately we need this to scale and we we will need DB functionality in any event. 

I think your outline provides the scope for an updated NB 04 for next week’s sprint.  Will await view from others to ensure they agree.

Perhaps we work on your item 1 and 3 for next week and design a mock DB for item 2.  We can either then add the real DB during the following week’s sprint or another developer can look at developing the DB API in parallel as a stand alone sprint task.  The reason I suggest this, is as you say, the DB will have additional functionality over and above user/password authentication, so needs to be scoped and thought through more broadly. 

What do others think?

Reply:
Broadly agree with what's been said here but I reckon we should use a mock DB api rather than setting up an actual DB. Will be the same architecture of accessing the data layer through an interface but using a mock DB means we don't have to worry about actually setting a DB up, as both options are rather different effort levels for the same reward.

Reply:
@ibrahimdirar That makes sense.

Reply:
I agree with you that this task should be coded in the proposed way. @MrDmitriev could you please add the tasks to the product backlog? I will prioritise the new tasks.